### DISCUSSION TITLE
Idea: Different input devices for different programs

### DISCUSSION THREAD

**Discussion by JackTheFragger** [nefarius/HidHide#60] [Ideas]
I want to have an option to configure hiding different input devices per program.

I have many input devices and some programs don't work with my xbox controller because they see to many controllers. So I would like to hide these other controllers for some programs. But these other controlers I need for other programs. So it would be very helpful to have "profiles" or something to define different devices to hide for different programs.

**Comment by nefarius** [comment]
You can make your own profiles using the command line client.

**Reply by JackTheFragger** [reply]
There is no documentation for the command line client. I don't know how to create profiles with it. I've tried to use it from command shell with parameters but got access denied error:

Error code 0x0005 at src\FilterDriverProxy.cpp(31) `anonymous-namespace'::Device: - Access Denied

I've started Command Shell in admin mode.

Also I don't think that I can handle profiles with it because I have to write batch files to hide or unhide devices before starting a game. But most games are started by steam or epic launcher or similiar.

**Reply by nefarius** [reply]
The documentation is within the CLI help itself and we don't have the time or capacity to spend to write articles for it.

**Reply by nefarius** [reply]
You need to close the UI otherwise you get access denied. No Admin required.

**Comment by JackTheFragger** [comment]
I don't understand why you don't want to take up this idea. There are several similar suggestions on this. Nowadays most games are launched via Steam, Epic Launcher or similar. It would be a very big improvement for your software to have more rules for hiding devices. I looked at the code that hides the controls.

You look at the blacklist to see if a device is there. If it's there, look at the whitelist to see if the accessing process is there. If it is not present, access to the controller is denied. If you would change the behavior so that apps are blacklisted instead of whitelisted, you can have a separate controller blacklist for each app. So everyone could manage for each program which device should be hidden for this program.

I would prefer to have much finer control over which programs are allowed to see which controllers. And I think most other users would see that as an asset as well.

**Reply by nefarius** [reply]
Because... It's a ton of work and our plates are currently full and no capacity left so feel free to contribute and open a PR. Until then you gotta work with what you got. This is free software, nobody is owing anybody anything.

**Reply by nefarius** [reply]
Obligatory read before the next post with entitled undertone: https://mikemcquaid.com/2018/03/19/open-source-maintainers-owe-you-nothing/

**Comment by JackTheFragger** [comment]
Please don't misunderstand my intention. I know that nobody is owing me anything.

But accept an idea and implent it are to separate things. When you say you have no time - alright. No problem. You're answer suggested me that you are'nt interested in my or similiar ides. So I didn't understand it.

To my question of the CLI: I hadn't the UI opened. I can start the CLI. But I can't call it with parameters from the dos shell. So I can't use it for batches...

**Reply by nefarius** [reply]
> But accept an idea and implent it are to separate things. When you say you have no time - alright. No problem. You're answer suggested me that you are'nt interested in my or similiar ides. So I didn't understand it.

I very well understand and am interested in the idea but at the same time it makes no sense for me to further go down a rabbit hole with no achievable goal. Your suggestion would result in a complete redesign and most probably rewrite of the entire solution so the scope of work for this seemingly simple feature request is basically birthing a new project. The crew behind this particular project never intended such a spike in complexity nor has the resources for like the next 6 months or so to tackle this request so I don't see benefit in devoting time to speculating what *would* be a good feature. I don't just shrug off feature requests because I feel like it but at the same time I understand the work required to fulfil them and for this request it's not worth it and not gonna happen in the near future for the reasons I outlined.

> To my question of the CLI: I hadn't the UI opened. I can start the CLI. But I can't call it with parameters from the dos shell. So I can't use it for batches...

This then definitely sounds like a bug, it is absolutely intended for the CLI to work for batch scripts so I try to replicate this issue on my end. The only reason you get an access denied is when another process keeps a handle to the driver open at the same time, which is not allowed to mitigate concurrency issues.

> Please don't misunderstand my intention. I know that nobody is owing me anything.

Alright, seems we're on the same page after all ❤️

**Reply by JackTheFragger** [reply]
I'm using joystick gremlin which is whitelisted. Maybe it's keeps a handle to the driver open?

**Reply by nefarius** [reply]
Only apps with a (butchered) implementation of the HidHide API access it directly, so IDK if JSG does that, kill the process and try to invoke the CLI again as a simple test.

**Comment by JackTheFragger** [comment]
Hmm. Now my test is successful after closing Joystick Gremlin. But after starting Joystick Gremlin test are successful, too. Maybe there was another process which blocked it...

**Reply by nefarius** [reply]
A quick search through their source code doesn't bring up anything about HidHide so it probably wasn't JSG 🤔

**Reply by mrrw0lf** [reply]
were u able to create profiles i just typed help and i dont see any way how i can create anything profile like

**Reply by JackTheFragger** [reply]
It was HidHide itself. I had the GUI opened. After closing the CLI worked.

**Reply by mrrw0lf** [reply]
i have basically no idea of coding can u link me some sources and a short tldr on how to create those mentioned profiles for several Programms i basically want to hide a real controller from 2 Programms and and the emulated ds4 (from ds4 windows)only from one of the programs that the real one is hidden from

**Reply by JackTheFragger** [reply]
I react with scheduled task on events to start my scripts. I can send you my code for hiding and unhiding. You have to replace the IDs for your controllers. And you have to add or remove lines because of different amount of controllers I think.

**Reply by mrrw0lf** [reply]
uh u can but b4 we go through that hustle is youre code able to hide and unhide simultaneously or is it just alternating between profiles? as in having separate profiles for different hids running at the same time side by side

**Comment by JackTheFragger** [comment]
There are no profiles. I have scripts to hide and unhide controls. There is no possibility to hide different Controls to different programs.

**Reply by mrrw0lf** [reply]
too bad
